include/boost/geometry/algorithms/detail/for_each_range.hpp

```diff
@@ -38,47 +48,78 @@ namespace detail { namespace for_each
 {
 
 
-template <typename Range, typename Actor>
+template <typename Point>
+struct fe_range_point
+{
+    template <typename Functor>
+    static inline bool apply(Point& point, Functor&& f)
+    {
+        Point* ptr = boost::addressof(point);
+        return f(std::pair<Point*, Point*>(ptr, ptr + 1));
+    }
+};
+
+
+template <typename Segment>
+struct fe_range_segment
+{
+    template <typename Functor>
+    static inline bool apply(Segment& segment, Functor&& f)
+    {
+        return f(segment_view<typename boost::remove_const<Segment>::type>(segment));
+    }
+};
+
+
+template <typename Range>
 struct fe_range_range
 {
-    static inline void apply(Range & range, Actor & actor)
+    template <typename Functor>
+    static inline bool apply(Range& range, Functor&& f)
     {
-        actor.apply(range);
+        return f(range);
     }
 };
 
 
-template <typename Polygon, typename Actor>
+template <typename Polygon>
 struct fe_range_polygon
 {
-    static inline void apply(Polygon & polygon, Actor & actor)
+    template <typename Functor>
+    static inline bool apply(Polygon& polygon, Functor&& f)
     {
-        actor.apply(exterior_ring(polygon));
+        return f(exterior_ring(polygon));
 
         // TODO: If some flag says true, also do the inner rings.
         // for convex hull, it's not necessary
     }
 };
 
-template <typename Box, typename Actor>
+template <typename Box>
 struct fe_range_box
 {
-    static inline void apply(Box & box, Actor & actor)
+    template <typename Functor>
+    static inline bool apply(Box& box, Functor&& f)
     {
-        actor.apply(box_view<typename boost::remove_const<Box>::type>(box));
+        return f(box_view<typename boost::remove_const<Box>::type>(box));
     }
 };
 
-template <typename Multi, typename Actor, typename SinglePolicy>
+template <typename Multi, typename SinglePolicy>
 struct fe_range_multi
 {
-    static inline void apply(Multi & multi, Actor & actor)
+    template <typename Functor>
+    static inline bool apply(Multi& multi, Functor&& f)
     {
-        for ( typename boost::range_iterator<Multi>::type
-                it = boost::begin(multi); it != boost::end(multi); ++it)
+        auto const end = boost::end(multi);
+        for (auto it = boost::begin(multi); it != end; ++it)
```

Comment:
Since you're reworking this, it's nicer to use a range based loop here
`for (auto& geometry : multi)` {... `apply(geometry, f)` }

Reply:
This is probably not possible. It's because we require `MultiGeometry` to be adapted to Boost.Range concept. So it's possible to adapt it and at the same time avoid implementing `begin` and `end` members or free functions that could be found via ADL. Consider e.g. `std::pair<T*, T*>` which is adapted to Boost.Range concept but can't be used in range-based for loop.

Unless I'm missing something and e.g. Boost.Range defines some free functions enabled conditionally for adapted types. I quickly tried to find something like this but was unable to do it. Plus I'm not sure if they would implement something like this because it's not clear to me that this would not cause any problems.

Reply:
Actually we could replace it with something like:
```
for (auto& geometry : make_range(multi))
```
where
```
template <typename Range>
inline auto make_range(Range & rng)
{
    return boost::make_iterator_range(boost::begin(rng), boost::end(rng));
}
```